presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,50 @@ private PinotExpression handleLogicalBinary(
         }
         List<RowExpression> arguments = call.getArguments();
         if (arguments.size() == 2) {
-            return derived(format(
-                    "(%s %s %s)",
-                    arguments.get(0).accept(this, context).getDefinition(),
-                    operator,
-                    arguments.get(1).accept(this, context).getDefinition()));
+            String def0 = arguments.get(0).accept(this, context).getDefinition();
+            String def1 = arguments.get(1).accept(this, context).getDefinition();
+            def1 = handleTimeStampAndDateCast(arguments.get(0), def1);
+            def0 = handleTimeStampAndDateCast(arguments.get(1), def0);
+            return derived(format("(%s %s %s)", def0, operator, def1));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private String handleTimeStampAndDateCast(RowExpression rowExpression, String definition)
+    {
+        if (!(rowExpression instanceof CallExpression)) {
+            return definition;
+        }
+        CallExpression callExpr = (CallExpression) rowExpression;
+        if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
```

Comment:
I didn't follow why `handleTimeStampAndDateCast` is not called from `handleCast` ? I don't understand why the codepath is entering `handleLogicalBinary` vs `handleCast`.

Reply:
this is a bit tricky, the logic behind is that Presto can compare a timestamp to a Date literal or a date to a timestamp literal.
Cast only works when user explicitly convert the time field to corresponding literal type.
E.g. `WHERE eventDate < current_time`
We know that left is a date, right side is a timestamp.
but until hit the binary logic, we know that we need to convert current_time to date value so that the pushing down logic is correct with the value.